Replace mocked data in Top Charts with data from java servlet stream

essencec opened the pull request.

essencec committed:
Replace mocked data in Top Charts with data from java servlet stream

essencec committed:
Replace mocked data in Top Charts with data from java servlet stream

essencec commented:
![image](https://user-images.githubusercontent.com/45109455/87972482-f6a1f080-ca94-11ea-9278-bce550d2f967.png)

ajwang0 commented on client/src/Clusters/TopCharts/TopCharts.js:
```diff
@@ -18,20 +18,20 @@ class TopCharts extends Component {
-        {charts.map((value, index) => {
-            return <div className="col" key={index}>
+        {data.charts && data.charts.map((chart) => {
+            console.log(chart)
```
Remove log

ajwang0 commented on client/src/App.js:
```diff
@@ -8,22 +8,32 @@ import AppOfTheWeek from './Clusters/AppOfTheWeek/AppOfTheWeek.js';
+        {this.state.streamData.clusters.map((cluster) => {
+            switch(cluster.type) {
+            case "TopCharts":
+            return <TopCharts data={cluster} key={cluster.id} />
```
Can you indent the return so its easier to read

ajwang0 commented on src/main/java/com/google/sps/models/TopChartsCluster.java:
```diff
@@ -1,18 +1,24 @@
+public class TopChartsCluster extends Cluster {
     private String title = "Top Charts";
+    private String subtitle = "For Wear OS";
     private ArrayList<Chart> charts = new ArrayList<Chart>();
```
Can you also update this to use List 

`private List<Chart> charts = new ArrayList<Chart>();`

ajwang0 commented on src/main/java/com/google/sps/models/Cluster.java:
```diff
@@ -1,6 +1,7 @@
-    String getType();
-    void makeCluster();
+abstract class Cluster {
+    String type;
```
Can you add a new line inbetween String type and the methods

ajwang0 reviewed (changes requested):
I think you'll need to update the test for App.js to use the FetchMock like in https://github.com/googleinterns/play-web-step-2020/pull/36/files#diff-648495ee1d77c291870f467735fd9493

essencec committed:
Modify client tests

essencec committed:
Modify client tests

ajwang0 commented on src/main/java/com/google/sps/models/TopChartsCluster.java:
```diff
@@ -1,18 +1,24 @@
 package com.google.sps.servlets;
 import java.util.ArrayList;
```
Don't forget to also add `import java.util.List;`

ajwang0 commented on client/src/App.js:
```diff
@@ -8,22 +8,35 @@ import AppOfTheWeek from './Clusters/AppOfTheWeek/AppOfTheWeek.js';
-        .then((response) => response.text())
-        .then((text) => {this.setState({text})})
+        .then((response) => {
+            if(response.status == 200) {
```
Add space after if and use 3 equals

`if (response.status === 200) {`

tonymai commented on client/src/App.js:
```diff
@@ -8,22 +8,35 @@ import AppOfTheWeek from './Clusters/AppOfTheWeek/AppOfTheWeek.js';
   constructor(props) {
     super(props);
-    this.state = { text: '' };
+    this.state = { streamData: null };
```
`this.state = {}`

tonymai commented on client/src/App.js:
```diff
@@ -8,22 +8,35 @@ import AppOfTheWeek from './Clusters/AppOfTheWeek/AppOfTheWeek.js';
   }
 
   render() {
+    if (this.state.streamData == null) {
```
use triple equal signs, `===`

![image](https://user-images.githubusercontent.com/9425367/88091174-8bfbbe00-cb43-11ea-9ce4-08d5f0740f1c.png)

tonymai commented on client/src/Clusters/TopCharts/TopCharts.js:
```diff
@@ -18,20 +18,19 @@ class TopCharts extends Component {
-    const charts = [topFree, topGrossing, trending, topPaid];
-
+    const { data } = this.props;
+    
```
remove spaces

essencec committed:
Fix server tests

essencec committed:
Modify test in topCharts.test.js

ajwang0 commented on client/src/App.test.js:
```diff
@@ -1,21 +1,38 @@
+  render(
+      <FetchMock
+        options={{
+          matcher: '/api/v1/stream', method: 'GET', response: {"clusters":[{"title":"Top Charts","subtitle":"For Wear OS","charts":[{"title":"Top Free","apps":[{"id":"1","icon":"mockIcon1.png","title":"mockApp1","category":"Fake Apps","rating":5.0,"price":0.0},{"id":"2","icon":"mockIcon2.png","title":"mockApp2","category":"Fake Apps","rating":5.0,"price":0.0}]},{"title":"Top Grossing","apps":[{"i ...
```
Can you make a variable for the response and move the data into the var so this is cleaner to read?

```
const responseData = {"clusters": ...};

...
matcher: '/api/v1/stream', method: 'GET', response: responseData

```

ajwang0 commented on client/src/App.test.js:
```diff
@@ -1,21 +1,38 @@
+      </FetchMock>
+  );
+
+test('verifies fetch is called from Top Charts', async () => {
```
I would rename this to "fetch returns data for Top Charts cluster"

ajwang0 commented on client/src/App.test.js:
```diff
@@ -1,21 +1,38 @@
-    const { getByText } = render(<App />);
-    const textElement = getByText('App of the week');
-    expect(textElement).toBeInTheDocument();
+test('verifies nothing is rendered if streamData is set to null', async () => {
```
I would rename this to "nothing is rendered when fetch fails"

ajwang0 commented on client/src/Clusters/TopCharts/TopCharts.js:
```diff
@@ -18,20 +18,21 @@ class TopCharts extends Component {
-    const charts = [topFree, topGrossing, trending, topPaid];
-
+    const { data } = this.props;    
+    if (data == null) {
```
We can omit this since we only render TopCharts when App.js finds a TopCharts in the data.

ajwang0 commented on client/src/Clusters/TopCharts/TopCharts.js:
```diff
@@ -1,10 +1,10 @@
 import './TopChartsStyle.css';
 import '../styles.css';
  
 class TopCharts extends Component {
```
Can you also add PropTypes for this. At the minimum at least define data is a required object

ajwang0 commented on client/src/Clusters/TopCharts/topCharts.test.js:
```diff
@@ -1,9 +1,31 @@
+          data={data} />
+    );
+
+test('correct column titles', async() => {
```
Can you also verify the App cards are in the DOM?

essencec committed:
Create test to verify App cards

essencec commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogleinterns%2Fplay-web-step-2020%2Fpull%2F65) for more info**.

@googlebot I fixed it
